keel-core/src/main/kotlin/com/netflix/spinnaker/keel/services/ApplicationService.kt

```diff
@@ -85,16 +89,30 @@ class ApplicationService(
   fun pin(application: String, pin: EnvironmentArtifactPin, user: String) {
     val config = repository.getDeliveryConfigForApplication(application)
     repository.pinEnvironment(config, pin.copy(pinnedBy = user))
+    publisher.publishEvent(
+      ArtifactPinnedEvent(
+        application = application,
+        environment = pin.targetEnvironment,
+        timestamp = Instant.now(),
+        triggeredBy = user,
+        reference = pin.reference,
+        version = pin.version
+      )
+    )
   }
 
-  fun deletePin(application: String, pin: EnvironmentArtifactPin) {
-    val config = repository.getDeliveryConfigForApplication(application)
-    repository.deletePin(config, pin.targetEnvironment, pin.reference)
-  }
-
-  fun deletePin(application: String, targetEnvironment: String) {
+  fun deletePin(application: String, targetEnvironment: String, reference: String?, user: String) {
     val config = repository.getDeliveryConfigForApplication(application)
-    repository.deletePin(config, targetEnvironment)
+    repository.deletePin(config, targetEnvironment, reference)
+    publisher.publishEvent(
```

Comment:
We could previously figure out when something was pinned by looking at the database, but we didn't have any historical record of unpinning. This event adds that.

Reply:
Will open a separate PR for new events.